File: frontend/src/types/device.ts
import type { DeviceStatus } from "./enums";
import type { DepartmentResponse } from "./org";
import type { TagResponse } from "./org";
import type { UserResponse } from "./user";
import type { AlertResponse } from "./alert";
import type { LocationHistoryResponse } from "./location";

export interface DeviceResponse {
  id: string;
  name: string;
  model: string | null;
  manufacturer: string | null;
  androidVersion: string | null;
  imei: string | null;
  serialNumber: string | null;
  status: DeviceStatus;
  lastSyncAt: string | null;
  deviceOwnerActive: boolean;
  department: DepartmentResponse | null;
  responsibleUser: UserResponse | null;
  tags: TagResponse[];
  createdAt: string;
}

export interface CreateDeviceRequest {
  name: string;
  model?: string;
  manufacturer?: string;
  androidVersion?: string;
  imei?: string;
  serialNumber?: string;
  departmentId?: string;
  responsibleUserId?: string;
  tagIds?: string[];
}

export interface UpdateDeviceRequest {
  name: string;
  model?: string;
  manufacturer?: string;
  status: DeviceStatus;
  departmentId?: string;
  responsibleUserId?: string;
  tagIds?: string[];
}

export interface DeviceProvisionResponse {
  deviceId: string;
  apiKey: string;
}

export interface DashboardSummaryResponse {
  totalDevices: number;
  onlineDevices: number;
  offlineDevices: number;
  lowBatteryDevices: number;
  unreadAlerts: number;
  recentAlerts: AlertResponse[];
  recentLocationsByDevice: Record<string, LocationHistoryResponse>;
}

File: backend/src/main/java/com/tactio/mdm/domain/entity/Device.java
package com.tactio.mdm.domain.entity;

import com.tactio.mdm.domain.enums.DeviceStatus;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;

import java.time.Instant;
import java.util.HashSet;
import java.util.Set;

@Getter
@Setter
@EqualsAndHashCode(callSuper = true)
@Entity
@Table(name = "devices")
public class Device extends BaseEntity {

    @Column(nullable = false, length = 150)
    private String name;

    @Column(length = 100)
    private String model;

    @Column(length = 100)
    private String manufacturer;

    @Column(length = 20)
    private String androidVersion;

    @Column(length = 30, unique = true)
    private String imei;

    @Column(length = 60, unique = true)
    private String serialNumber;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private DeviceStatus status = DeviceStatus.PROVISIONING;

    private Instant lastSyncAt;

    @Column(nullable = false)
    private boolean deviceOwnerActive = false;

    @Column(length = 500)
    private String fcmToken;

    @Column(length = 255)
    private String apiKeyHash;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "department_id")
    private Department department;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "responsible_user_id")
    private User responsibleUser;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "device_tags",
            joinColumns = @JoinColumn(name = "device_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    private Set<Tag> tags = new HashSet<>();
}
